File: 20.03.2023/js/main.js
// Given a string with the weights of FFC members in normal order,
// give this string ordered by "weights" of these numbers


/* Solution one */
export const orderWeightOne = (str) => {
    // Remove leading and trailing whitespaces and split the string into an array
    const arr = str.trim().split(/\s+/);
    
    // Define a function to calculate the weight of a number
    function weight(num) {
        return num.split('').reduce((sum, digit) => sum + parseInt(digit), 0);
    }
    
    // Sort the array of numbers by their weight and then by their string value
    arr.sort((a, b) => {
        const weightA = weight(a);
        const weightB = weight(b);

        if (weightA !== weightB) {
            return weightA - weightB;
        } else {
            return a.localeCompare(b);
        }
    });
    
    // Join the sorted array into a string with a single space between numbers
    return arr.join(' ');
}

/* Solution two */
export const orderWeightTwo = (str) => {
    // Remove leading and trailing whitespaces and split the string into an array
    const arr = str.trim().split(/\s+/);
    
    // Define a function to calculate the weight of a number
    function weight(num) {
        return num.split('').reduce((sum, digit) => sum + parseInt(digit), 0);
    }
    
    // Map each number to an object with its original value and its weight
    const weightArr = arr.map(num => ({ value: num, weight: weight(num) }));
    
    // Sort the array of objects by weight and then by string value
    weightArr.sort((a, b) => {
        if (a.weight !== b.weight) {
            return a.weight - b.weight;
        } else {
            return a.value.localeCompare(b.value);
        }
    });
    
    // Map the sorted array of objects back to an array of values
    const sortedArr = weightArr.map(obj => obj.value);
    
    // Join the sorted array into a string with a single space between numbers
    return sortedArr.join(' ');
}

/* Solution three */
export const orderWeightThree = (str) => {
    // Remove leading and trailing whitespaces and split the string into an array
    const arr = str.trim().split(/\s+/);
    
    // Define a function to calculate the weight of a number
    function weight(num) {
        return num.split('').reduce((sum, digit) => sum + parseInt(digit), 0);
    }
    
    // Reduce the array of numbers to an object with keys as weights and values as arrays of numbers
    const weightObj = arr.reduce((obj, num) => {
        const w = weight(num);
        obj[w] = obj[w] || [];
        obj[w].push(num);
        return obj;
    }, {});
    
    // Sort the arrays of numbers by string value
    Object.values(weightObj).forEach(arr => arr.sort((a, b) => a.localeCompare(b)));
    
    // Sort the keys of the object numerically
    const sortedWeights = Object.keys(weightObj).sort((a, b) => a - b);
    
    // Map the sorted arrays of numbers to a single array and join into a string
    const sortedArr = sortedWeights.flatMap(weight => weightObj[weight]);

    return sortedArr.join(' ');
}

File: 20.03.2023/py/main.py
# Given a string with the weights of FFC members in normal order,
# give this string ordered by "weights" of these numbers


class WrongType(Exception):
    pass

# Solution one
def weight(num):
    return sum(int(digit) for digit in str(num))

def order_weight_one(strng):
    nums = strng.split()
    nums.sort(key=lambda num: (weight(num), num))

    return ' '.join(nums)

# Solution two
def order_weight_two(strng):
    nums = strng.split()
    weight_dict = {num: sum(int(digit) for digit in num) for num in nums}
    sorted_nums = sorted(nums, key=lambda num: (weight_dict[num], num))

    return ' '.join(sorted_nums)

# Solution three
def order_weight_three(strng):
    nums = strng.split()

    def weight(num):
        return sum(int(digit) for digit in num)

    sorted_nums = sorted(nums, key=lambda num: (weight(num), num))

    return ' '.join(sorted_nums)

# Solution four
def order_weight_four(strng):
    nums = strng.split()
    weighted_nums = [(num, sum(int(digit) for digit in num)) for num in nums]
    sorted_nums = [num[0] for num in sorted(weighted_nums, key=lambda num: (num[1], num[0]))]

    return ' '.join(sorted_nums)


import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('order_weight_one("103 123 4444 99 2000")')
    cProfile.run('order_weight_two("103 123 4444 99 2000")')
    cProfile.run('order_weight_three("103 123 4444 99 2000")')
    cProfile.run('order_weight_four("103 123 4444 99 2000")')

    start_time = time.time()
    order_weight_one('103 123 4444 99 2000')
    order_weight_two('103 123 4444 99 2000')
    order_weight_three('103 123 4444 99 2000')
    order_weight_four('103 123 4444 99 2000')
    end_time = time.time()

    print('', end_time - start_time)
